pallets/verifiers/src/lib.rs

```diff
@@ -174,31 +175,56 @@ pub mod pallet {
         }
     }
 
-    fn statement_hash(ctx: &[u8], vk_hash: &H256, version_hash: Option<H256>, pubs: &[u8]) -> H256 {
+    /// Compute the statement hash for a given vk, proof, and public data.
+    pub fn compute_statement_hash<I: Verifier>(
+        vk_or_hash: &VkOrHash<I::Vk>,
+        proof: &I::Proof,
+        pubs: &I::Pubs,
+    ) -> H256 {
+        let version_hash = I::verifier_version_hash(proof);
+        let hash = match vk_or_hash {
+            VkOrHash::Hash(h) => sp_std::borrow::Cow::Borrowed(h),
+            VkOrHash::Vk(vk) => sp_std::borrow::Cow::Owned(I::vk_hash(vk)),
+        };
+        let ctx: &[u8] = I::hash_context_data();
+        let vk_hash: &H256 = hash.as_ref();
+        let pubs = I::pubs_bytes(pubs);
+
         let mut data_to_hash = keccak_256(ctx).to_vec();
         data_to_hash.extend_from_slice(vk_hash.as_bytes());
         if let Some(h) = version_hash {
             data_to_hash.extend_from_slice(h.as_bytes());
         }
-        data_to_hash.extend_from_slice(keccak_256(pubs).as_bytes_ref());
+        data_to_hash.extend_from_slice(keccak_256(pubs.as_bytes_ref()).as_bytes_ref());
         H256(keccak_256(data_to_hash.as_slice()))
     }
 
-    fn compute_hash<I: Verifier>(
-        version_hash: Option<H256>,
-        pubs: &I::Pubs,
+    pub(crate) fn submit_proof_weight<T: Config<I>, I: 'static + Verifier>(
         vk_or_hash: &VkOrHash<I::Vk>,
-    ) -> H256 {
-        let hash = match vk_or_hash {
-            VkOrHash::Hash(h) => sp_std::borrow::Cow::Borrowed(h),
-            VkOrHash::Vk(vk) => sp_std::borrow::Cow::Owned(I::vk_hash(vk)),
+        proof: &I::Proof,
+        pubs: &I::Pubs,
+        domain_id: &Option<u32>,
+    ) -> Weight {
+        // Check disable: we din't consider any time cost about checking boolean
+        // variable and proof size: we consider them negligible
+        let base = T::DbWeight::get().reads(1);
+        let vk_weight = match vk_or_hash {
+            VkOrHash::Hash(_) => {
+                // We considering unwrapping VkEntry negligible
+                T::WeightInfo::get_vk()
+            }
+            VkOrHash::Vk(_) => {
+                // We considering cloning vk negligible
+                T::WeightInfo::validate_vk(None, Some(proof), Some(pubs))
+            }
```

Comment:
Here you have access to the actual vk

```suggestion
            VkOrHash::Vk(vk) => {
                // We considering cloning vk negligible
                T::WeightInfo::validate_vk(Some(vk), None, None)
            }
```

If you do like this, then you can switch the `hp_verifiers::WeightInfo::validate_vk` signature from
```rust
fn validate_vk(vk: Option<&V::Vk>, proof: Option<&V::Proof>, pubs: Option<&V::Pubs>) -> Weight;
```
to the simpler
```rust
fn validate_vk(vk: &V::Vk) -> Weight;
```

Reply:
Thanks again!!, I fixed it